src/main/java/io/supertokens/oauth/OAuth.java

```diff
@@ -66,39 +101,181 @@ public static OAuthAuthResponse getAuthorizationUrl(Main main, AppIdentifier app
+            throw new IllegalStateException("Unexpected answer from Oauth Provider");
+        }
+        if(responseHeaders.containsKey(COOKIES_HEADER_NAME)){
+            cookies = new ArrayList<>(responseHeaders.get(COOKIES_HEADER_NAME));
 
-                if(Utils.containsUrl(locationHeaderValue, hydraBaseUrlForConsentAndLogin, true)){
-                    redirectTo = locationHeaderValue.replace(hydraBaseUrlForConsentAndLogin, "{apiDomain}");
-                } else {
-                    redirectTo = locationHeaderValue;
+            for (int i = 0; i < cookies.size(); i++) {
+                String cookieStr = cookies.get(i);
+                if (cookieStr.startsWith("ory_hydra_")) {
+                    cookies.set(i, "st_oauth_" + cookieStr.substring(10));
                 }
+            }
+        }
+
+        return new OAuthAuthResponse(redirectTo, cookies);
+    }
+
+    public static JsonObject getToken(Main main, AppIdentifier appIdentifier, Storage storage, JsonObject bodyFromSdk, String iss, boolean useDynamicKey) throws InvalidConfigException, TenantOrAppNotFoundException, OAuthAPIException, StorageQueryException, IOException, InvalidKeyException, NoSuchAlgorithmException, InvalidKeySpecException, JWTCreationException, JWTException, StorageTransactionLogicException, UnsupportedJWTSigningAlgorithmException, FeatureNotEnabledException {
+        checkForOauthFeature(appIdentifier, main);
+        
+        OAuthStorage oauthStorage = StorageUtils.getOAuthStorage(storage);
+        String clientId = bodyFromSdk.get("client_id").getAsString();
+
+        if (!oauthStorage.doesClientIdExistForThisApp(appIdentifier, clientId)) {
+            throw new OAuthAPIException("invalid_client", "Client authentication failed (e.g., unknown client, no client authentication included, or unsupported authentication method). The requested OAuth 2.0 Client does not exist.", 400);
+        }
+
+        String publicOAuthProviderServiceUrl = Config.getConfig(appIdentifier.getAsPublicTenantIdentifier(), main).getOAuthProviderPublicServiceUrl();
+        try {
+            Map<String, String> bodyParams = constructHydraRequestParamsForAuthorizationGETAPICall(bodyFromSdk);
+            if (bodyParams.containsKey("code")) {
+                bodyParams.put("code", bodyParams.get("code").replace("st_ac_", "ory_ac_"));
+            }
+            if (bodyParams.containsKey("refresh_token")) {
+                bodyParams.put("refresh_token", bodyParams.get("refresh_token").replace("st_rt_", "ory_rt_"));
+            }
+            JsonObject response = HttpRequest.sendFormPOSTRequest(main, "", publicOAuthProviderServiceUrl + HYDRA_TOKEN_ENDPOINT, bodyParams, 10000, 10000, null);
+
+            // token transformations
+            if (response.has("access_token")) {
+                String accessToken = response.get("access_token").getAsString();
+                accessToken = reSignToken(appIdentifier, main, accessToken, iss, 1, useDynamicKey, 0);
```

Comment:
The `stt` values should probably go into an enum/separate const.

Reply:
Done